### PR TITLE
CDAP-15785 remove redundant check for fll

### PR DESCRIPTION
**Pull request by yaojiefeng**
JIRA: https://issues.cask.co/browse/CDAP-15785
build: https://builds.cask.co/browse/CDAP-RUT1725-1

The check for the redundant operations is too strict and it is very easy for user to make a mistake and get the entire lineage graph invisible. Instead of failing the lineage check, change this to a log message.

### CODE REVIEW

**Comment by nitinmotgi** [comment]
@mikkin this is needed for lineage to work.

**Review by mikkin — APPROVED** [review]

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/FieldLineageProcessor.java [hunk 1]
```diff
@@ -67,6 +67,7 @@ public FieldLineageProcessor(PipelineSpec pipelineSpec) {
 
     // validate the stage operations
     Map<String, InvalidFieldOperations> stageInvalids = new HashMap<>();
+    Map<String, Map<String, List<String>>> stageRedundants = new HashMap<>();
```
Comment (CuriousVini) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/FieldLineageProcessor.java:70]: should we use another class similar to `InvalidFieldOperations` to follow the same pattern? This seem to be exposing the ds directly.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/FieldLineageProcessor.java [hunk 2]
```diff
@@ -124,13 +125,20 @@ public FieldLineageProcessor(PipelineSpec pipelineSpec) {
...
+
+    if (!stageRedundants.isEmpty()) {
+      LOG.debug("The pipeline has redundant operations {} and they will be ignored", stageRedundants);
     }
 
     if (!stageInvalids.isEmpty()) {
       // Do not throw but just log the exception message for validation failure
       // Once most of the plugins are updated to write lineage exception can be thrown
       LOG.debug(new InvalidLineageException(stageInvalids).getMessage());
-      return Collections.emptySet();
```
Comment (CuriousVini) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/FieldLineageProcessor.java:133]: why is this change needed?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/StageOperationsValidator.java [hunk 3]
```diff
@@ -163,34 +171,23 @@ void validate() {
...
+      // decides to drop the joiner key in the output schema. The operation
+      // [D1.K1, D2.K2] ->[K1, K2] is a valid even though K1,K2 are not in the output schema.
+      iterator.remove();
     }
 
     this.invalidOutputs.putAll(unusedOutputs.entrySet().stream().collect(
       Collectors.toMap(Map.Entry::getKey,
                        e -> e.getValue().stream().map(FieldOperation::getName).collect(Collectors.toList()))));
-    this.invalidOutputs.putAll(redundantOutputs.entrySet().stream().collect(
+    this.redundantOutputs.putAll(redundantOutputs.entrySet().stream().collect(
```
Comment (CuriousVini) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/StageOperationsValidator.java:190]: would it be a good idea to create another method for this since it uses exactly same logic?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/FieldLineageProcessor.java (hunk 2 above)
Comment (yaojiefeng) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/FieldLineageProcessor.java:133]: This is to make sure we will process the rest field lineage. If we return an empty set here, the lineage will be gone. We want to make sure the valid field lineage is written.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/StageOperationsValidator.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/StageOperationsValidator.java:190]: This is just a one line which does the convert. I don't think there is a need to create another method for it. And we may revisit the validation logic later since the current validation is too strict so the these fields might get changed in the future.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/FieldLineageProcessor.java (hunk 1 above)
Comment (yaojiefeng) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/FieldLineageProcessor.java:70]: `InvalidFieldOperations` is just a wrapper of two `Map<String, List<String>>` of input and output fields. Here we don't need that as we only need one map. The ds will not get exposed. The map is like <stageName, <field, List<operation name>>>, which is basically is same as `InvalidFieldOperations`

**Review by CuriousVini — APPROVED** [review]
